[Spring 지하철 노선도 - 1,2단계] 정(정진우) 미션 제출합니다.

bugoverdose opened the pull request:
안녕하세요. 정이라고 합니다.
뭔가 너무 빨리 끝나버려서 이렇게 하는게 맞나 싶네요.

1. Service layer 관련 
- 현재 요구사항처럼 데이터에 대한 간단한 CRUD 작업만 존재하는 경우, 컨트롤러와 Dao 사이에 굳이 별도의 layer가 존재할 필요성이 적다고 느꼈습니다. 
- 그래도 비즈니스 로직이 늘어나게 되면 결국에는 서비스가 필요해질 것 같아 일단은 만들었습니다.

2. `schema.sql`을 사용할 경우, 모든 테스트들이 하나의 test db에 종속될 것 같아, `@Sql`을 활용하여 테스트 관련 셋업/초기화 작업을 수행했습니다.

3. Dao에서는 `NamedParameterJdbcTemplate`를 활용하였고, 각종 예외를 메서드 체이닝으로 처리하기 위해 별도의 객체를 몇 개 만들었습니다.

bugoverdose committed:
docs(README): add feature list for step 1

bugoverdose committed:
feat(station): implement API for station

- implement StationDao and StationsController
- add test cases

bugoverdose committed:
feat(exception): implement ExceptionAdvice for exception handling

- implement custom exception type for 404 NOT FOUND

bugoverdose committed:
feat(Line): implement Line domain and LineDao

bugoverdose committed:
feat(line): implement API for line

- implement LineDao and LinesController
- add test cases

bugoverdose committed:
refactor(subway): minor cleansing

bugoverdose committed:
feat(config): setup db configuration

bugoverdose committed:
feat(dao): migrate StationDao to Spring JDBC

- add test cases

bugoverdose committed:
feat(dao): migrate LineDao to Spring JDBC

- add test cases

bugoverdose committed:
test: setup different test db for acceptance tests and dao unit tests

bugoverdose committed:
feat(controller): migrate to Spring JDBC daos

bugoverdose committed:
refactor(dao): remove legacy dao package

- replace with new daos using Spring JDBC

bugoverdose committed:
refactor(dto): cleanse and format all dtos

- delete unused dtos
- create sub packages to separate request and response dto
- add default constructor and setters for request dtos
- override toString methods

bugoverdose committed:
feat(service): implement LineService and StationsService

bugoverdose committed:
refactor(controller): add service layer between controller and dao

bugoverdose committed:
refactor(subway): minor cleansing

bugoverdose committed:
feat(dao): implement StatementExecutor and UpdateResult to cleanse exception handling

bugoverdose committed:
refactor(dao): cleanse logic with StatementExecutor and UpdateResult

- replace try-catch with method chaining

bugoverdose committed:
refactor(test): cleanse all tests

jjj0611 commented on src/main/java/wooteco/subway/dao/LineDao.java:
```diff
@@ -0,0 +1,79 @@
+    private static final String LINE_NOT_FOUND_EXCEPTION_MESSAGE = "해당되는 노선은 존재하지 않습니다.";
+    private static final String DUPLICATE_NAME_EXCEPTION_MESSAGE = "중복되는 이름의 지하철 노선이 존재합니다.";
+
+    private final NamedParameterJdbcTemplate jdbcTemplate;
```
이 친구는 언제 생성이 되고, 어떻게 여기에 주입되게 되나요??

bugoverdose replied:
JdbcTemplate과 NamedParameterJdbcTemplate의 경우 코드에 명시적으로 존재하지 않든, 존재하지 않든, 스프링 부트의 IoC 컨테이너에서 (의존성들 다 넣어서) 디폴트로 인스턴스화해주는 것으로 보여요.

그리고 LineDao의 생성자에 NamedParameterJdbcTemplate 타입의 매개변수가 있기 때문에 IoC 컨테이너에서는 해당 빈을 의존성으로 자동으로 주입해주는 것으로 압니다.

kang-hyungu commented on src/main/java/wooteco/subway/WebConfig.java:
```diff
@@ -6,8 +6,11 @@
+                .allowedMethods("*")
+                .allowedOriginPatterns("*");
     }
 }
```
깃헙에서 왜 빨간 원을 띄웠을까요?

bugoverdose replied:
IntelliJ 설정에서 모든 파일이 빈 줄 하나로 끝나도록 설정해서 일괄 수정했습니다.

<img width="391" alt="image" src="https://user-images.githubusercontent.com/73531614/167079502-0e079f71-0250-497e-a948-aa9586d69362.png">

bugoverdose replied:
근데 빨간 원이라는게 Files Changed 탭에서만 보이고 리뷰 댓글에는 안보이는 것 같아요

kang-hyungu commented on src/main/java/wooteco/subway/dao/LineDao.java:
```diff
@@ -0,0 +1,79 @@
+        final String sql = "SELECT * FROM line";
+
+        List<Line> lines = jdbcTemplate.query(sql, new EmptySqlParameterSource(), lineRowMapper);
+        return Collections.unmodifiableList(lines);
```
Collections.unmodifiableList로 감사줘야 할 이유가 있을까요?

bugoverdose replied:
음 가독성 차원에서 조회 목적으로 사용된다는 점을 명시하려고 했던 것으로 기억해요. 현재 로직에서는 사실 있든 없든 차이가 없다고 생각해요.

kang-hyungu commented on src/main/java/wooteco/subway/dao/LineDao.java:
```diff
@@ -0,0 +1,79 @@
+                .throwOnNonEffected(() -> new IllegalArgumentException(LINE_NOT_FOUND_EXCEPTION_MESSAGE));
+    }
+
+    private final RowMapper<Line> lineRowMapper = (resultSet, rowNum) ->
```
인스턴스 변수는 어느 위치에 작성하는게 좋을까요?

bugoverdose replied:
인스턴스 변수라고 생각하자니 거부감이 들어서 static 붙이고 위로 올렸습니다

kang-hyungu commented on src/main/java/wooteco/subway/service/LineService.java:
```diff
@@ -0,0 +1,46 @@
+        lineDao.deleteById(id);
+    }
+
+    private LineResponse toLineResponseDto(Line line) {
```
```suggestion
    private LineResponse toLineResponse(Line line) {
```
dto 역할을 메서드명에 표시해줄 필요는 없어보이네요.

kang-hyungu commented on src/main/java/wooteco/subway/ui/ExceptionAdvice.java:
```diff
@@ -0,0 +1,27 @@
+
+    @ExceptionHandler
+    public ResponseEntity<String> handleException(Exception e) {
+        e.printStackTrace();
```
적절한 예외 처리일까요?
콘솔 화면에서만 예외 메시지를 봐도 충분할까요?

bugoverdose replied:
`ResponseEntity<String>`로 예외 메시지를 응답에 포함시켜서 반환하고는 있는데, 다른 걸 의미하신건가요?

kang-hyungu commented on src/main/java/wooteco/subway/ui/LineController.java:
```diff
@@ -0,0 +1,63 @@
+    @PutMapping("/{id}")
+    public ResponseEntity<Void> updateLine(@PathVariable Long id,
+                                           @RequestBody LineRequest lineRequest) {
+        Line line = new Line(id, lineRequest.getName(), lineRequest.getColor());
```
1. 컨트롤러에서 특정 로직을 수행하지 말 것
2. 사용자가 정상적인 요청 값을 보냈을 거라 단정하지 말 것. id로 라인 조회해서 중복 체크 및 올바른 값인지 검증하고 수정할 것

bugoverdose replied:
컨트롤러와 DAO 사이에 서비스를 추가하는 과정에서 놓친게 좀 있는 것 같아요.

예외 처리의 경우 DAO에서 DB에 보낸 요청의 결과에 따라 일단 왠만한 경우들에 대해 예외를 던지고 있는데요.

현재는 예를 들어, 특정 id에 해당되는 역을 제거하라는 요청이 들어오면, (1) DAO에서 곧바로 "DELETE FROM station WHERE id = :id"라는 sql문을 DB로 보내고, (2) 수정된 로우가 0개라면, id에 해당되는 데이터는 없을 것이라고 간주하고, 예외 발생시켜서, (3) 최종적으로 "해당되는 역은 존재하지 않습니다."라는 메시지와 함께 404로 응답하는 구조인데요.

(1) id에 해당되는 데이터를 사전에 조회를 하고, (2) 서비스에서 검증 완료한 후, (3) DAO에서 실제로 데이터를 변경시키는 구조로 수정해야 한다는 것 맞나요?

kang-hyungu commented on src/test/java/wooteco/subway/acceptance/LineAcceptanceTest.java:
```diff
@@ -0,0 +1,228 @@
+                put("color", "bg-red-600");
+            }};
+
+            ExtractableResponse<Response> response = RestAssured.given().log().all()
```
RestAssured를 활용한 http 호출이 중복해서 나타나고 있네요.
중복을 제거할 방법 없을까요?

bugoverdose replied:
test 패키지에 유틸성 클래스를 만들어서 리팩토링해보았습니다.

kang-hyungu reviewed (changes requested):
정 몇 가지 피드백 남겼어요.
수정해보시고 리뷰 재요청주세요.
궁금한 점 생기면 코멘트 남기시거나 오프라인으로 찾아오셔도 좋습니다.

bugoverdose committed:
refactor(test): implement util classes for building and sending http request

- cleanse verbose and duplicate logic

bugoverdose committed:
refactor(test): change HttpRequestMessage to HttpUtils

- fix all use cases
- minor cleansing

bugoverdose committed:
style: make every file end with a single empty line

- configure intelliJ settings

bugoverdose committed:
refactor(dao): add static to row mappers

- change row mapper position

bugoverdose committed:
refactor(controller): remove model/entity from controller

- send dto to service

bugoverdose committed:
refactor(service): rename private method toLineResponseDto

bugoverdose committed:
refactor(controller): remove domain from StationController

- rename StationsService to StationService

jjj0611 commented on src/main/java/wooteco/subway/WebConfig.java:
```diff
@@ -6,8 +6,11 @@
     }
-}
\ No newline at end of file
+}
```
구구의 코멘트에 이어 EOF를 검색해보시겠어요?

bugoverdose replied:
운영체제에 따라 파일의 맨 끝이 공백으로 끝나야만 파일의 마지막 줄로 간주하는 경우가 있다고 들었습니다.
그 문제가 아니더라도 기본적으로 한줄 공백을 두는 것이 컨벤션인 것으로 압니다.

jjj0611 commented on src/main/java/wooteco/subway/dao/LineDao.java:
```diff
@@ -0,0 +1,79 @@
+    private static final String LINE_NOT_FOUND_EXCEPTION_MESSAGE = "해당되는 노선은 존재하지 않습니다.";
+    private static final String DUPLICATE_NAME_EXCEPTION_MESSAGE = "중복되는 이름의 지하철 노선이 존재합니다.";
+
+    private static final RowMapper<Line> lineRowMapper = (resultSet, rowNum) ->
```
constant의 컨벤션에 맞는걸까요?

bugoverdose replied:
람다를 저렇게 관리한 적은 딱히 없어서 확신이 없네요.

jjj0611 replied:
엇 정 네이밍 컨벤션에 대한 이야기를 드린거였어요 :sweat_smile:
기본적인 컨벤션을 지켜주세요!

jjj0611 commented on src/main/java/wooteco/subway/dao/LineDao.java:
```diff
@@ -0,0 +1,79 @@
+        SqlParameterSource paramSource = new BeanPropertySqlParameterSource(line);
+
+        new StatementExecutor<>(() -> jdbcTemplate.update(sql, paramSource, keyHolder))
+                .executeOrThrow(() -> new IllegalArgumentException(DUPLICATE_NAME_EXCEPTION_MESSAGE));
```
모든 DataAccessException 상황에서 항상 DUPLICATE_NAME을 가지고 예외를 던지게 되는거 아닐까요?
적절한 처리 방법일까요?

bugoverdose replied:
그게 제가 위에서 한 질문이고, 로컬에서 수정하고 있다고 한 부분이에요.

이걸 해결하려면 매번 별도로 요청을 보내서 특정 이름에 해당되는 데이터가 존재하는지를 먼저 체크해줘야 하는데, 
현재 전 애초에 그렇게 동작하는 것이 옳은가?에 대한 근본적인 의문을 지니고 있어요.

jjj0611 replied:
> ex) "데이터 저장을 시도했는데 실패했다면, name 컬럼에 대해 설정해놓은 unique constraint 때문일 것이다."

> ex) "데이터 수정/삭제를 시도했는데 예외는 발생하지 않았지만, 수정된 컬럼의 개수가 0개라면, id에 해당되는 데이터가 존재하지 않기 때문일 것이다."

이건 추측 아닌가요? 추측하지 말아주세요.

그리고 제가 한 질문은 적절한 처리 방법이 아니라는걸 말씀드린거에요.

> 이걸 해결하려면 매번 별도로 요청을 보내서 특정 이름에 해당되는 데이터가 존재하는지를 먼저 체크해줘야 하는데,

이게 맞는 가정인가요? 별도로 요청을 보내는 경우 이걸 해결하는 유일한 방법일까요?

bugoverdose replied:
네. 저도 저렇게 추측하는 방식이 부적절한 것 같아서 수정했습니다.
또 다른 이유는 "삭제할 데이터가 없다"와 같은 상황에 따른 예외 처리는 데이터 접근 작업보다는 비즈니스 로직과 관련된 책임인 것 같다는 생각이 들어서 DAO가 아닌 서비스로 예외처리를 옮겼습니다. 

흠. (1) SELECT문을 수행한 후에 UPDATE/DELETE 등을 수행하는 방식 외에는 (2) 애플리케이션 내에서 존재하는 station들의 이름과 id값 등의 목록을 관리하는 정도밖에 안떠오르네요. 

단순히 delete의 결과에 따른 effected row가 0인지 1인지를 참고하는 경우, 어떠한 이유에서 effected row의 값이 그러한지를 추측해줘야 하기 때문에 다시 원점으로 돌아오게 되는데, 그게 아니면 delete에서 다른 결과값을 반환하도록 할 수 있을까요?

jjj0611 replied:
> 어떠한 이유에서 effected row의 값이 그러한지를 추측해줘야 하기 때문에 다시 원점으로 돌아오게 되는데

이걸 꼭 추측해야만 할까요?? 추측하지 않고 그런 상황이 맞는지 확인하면 되죠!
위에 정이 들었던 예시에서 첫번째 예시는 완전한 추측이지만,
두번째 예시는 `예외는 발생하지 않았지만`이라는 조건이 있다면, 단순한 추측은 아닐 것 같아요

추측하지 않고, 근거를 세우고 검증해보면 어떨까요? :)

bugoverdose replied:
음 넵. 확실한건 제가 DB에 요청을 많이 보내는 거에 거부감이 있어서 사고가 경직된 것 같다는 생각이 드네요.

jjj0611 commented on src/main/java/wooteco/subway/dao/LineDao.java:
```diff
@@ -0,0 +1,79 @@
+        paramSource.addValue("id", id);
+
+        return new StatementExecutor<>(() -> jdbcTemplate.queryForObject(sql, paramSource, lineRowMapper))
+                .executeOrThrow(() -> new NotFoundException(LINE_NOT_FOUND_EXCEPTION_MESSAGE));
```
ditto. 아래 리뷰를 참고해주세요

jjj0611 commented on src/main/java/wooteco/subway/dao/StationDao.java:
```diff
@@ -1,30 +1,59 @@
+    private static final String STATION_NOT_FOUND_EXCEPTION_MESSAGE = "해당되는 역은 존재하지 않습니다.";
+    private static final String NAME_NOT_ALLOWED_EXCEPTION_MESSAGE = "해당 이름의 지하철역을 생성할 수 없습니다.";
+
+    private static final RowMapper<Station> stationRowMapper = (resultSet, rowNum) ->
```
ditto 상수의 컨벤션이 적절한가요??

jjj0611 commented on src/main/java/wooteco/subway/dao/StatementExecutor.java:
```diff
@@ -0,0 +1,33 @@
+import java.util.function.Supplier;
+import org.springframework.dao.DataAccessException;
+
+public class StatementExecutor<T> {
```
이 Executor는 혹시 어떤 것에서 영감을 받아서 만드셨을까요??

bugoverdose replied:
그냥 레벨1 체스 미션 때, 스프링 없이 jdbc를 직접 다룰 때 가급적 메서드 체이닝으로 DB에 요청을 보내고 싶다는 의도로 구현했었어요.

jjj0611 replied:
> 가급적 메서드 체이닝으로 DB에 요청을 보내고 싶다는 의도

정이 생각한 이것의 효용은 무엇일까요??

bugoverdose replied:
클라이언트 코드의 가독성이 굉장히 높아진다고 생각했어요. 흔히 말하는 "how가 아닌 what 중심의 코드"를 작성할 수 있었어요.

jjj0611 commented on src/main/java/wooteco/subway/domain/Line.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    @Override
+    public boolean equals(Object o) {
```
id가 같아도 name이나 color가 다르면 다른 line이 되겠군요!

bugoverdose replied:
일일이 getter를 호출하지 않아도 테스트할 수 있도록 일부러 저렇게 오버라이드를 한거였는데, 
id에 대해서만 equals/hashcode를 구현하는 것이 효용이 있는 경우가 혹시 있을까요?

```java
Line actual = dao.save(new Line("새로운 노선", "분홍색"));
Line expected = new Line(4L, "새로운 노선", "분홍색");
assertThat(actual).isEqualTo(expected);
```

bugoverdose replied:
일단 id로 식별하는건 맞는것 같아서 수정했습니다.

jjj0611 replied:
> 일단 id로 식별하는건 맞는 것 같아서

맞는 것 같다고 판단한 정의 근거는 무엇인가요?

bugoverdose replied:
음. 이 점에 대해서는 아직 확신이 없는데요. 

비즈니스 로직에 사용되는 domain이라면 모든 필드의 값을 사용해서 동등성 비교를 하는 것이 맞다고 생각해요.

다만, DB의 각 컬럼에 대응되는 정보를 담고 있는 entity인 경우 애매하다고 생각해요. 
(1) DB에서는 id 값을 데이터에 대한 식별자로 사용하므로 **id가 같은 객체들은 전부 동일한 대상을 지칭한다고 보는 것이 맞는가?** 
(2) 아니면 동일한 대상이지만 **상태가 달라졌으므로 서로 별개의 대상으로 간주하는 것이 맞는가?**

논리적으로는 (1)이 맞는 것 같은데, 효용의 측면에서는 (1)에 어떠한 효용이 있는지 의문을 가지고 있었습니다.

jjj0611 commented on src/main/java/wooteco/subway/domain/Line.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    @Override
+    public String toString() {
```
사용하는 곳이 있을까요??

bugoverdose replied:
toString은 디버깅 용도로 모든 곳에서 사용되죠.

jjj0611 replied:
> 디버깅 용도로 `모든 곳`에서 사용된다

정말인가요?

bugoverdose replied:
모든 곳이라는 표현 대신 기본적으로 테스트할 때 사용된다고 표현해야할까요?

아니면 디버깅 이외의 용도가 있을까요? 일단 비즈니스 로직이나 뷰 목적으로 쓰면 안되는 것으로 알고 있습니다.

jjj0611 replied:
네 비즈니스 로직이나 뷰 목적으로 사용하지 않는게 좋습니다.

디버깅 용도로 사용하실 때는 사용하는 곳에서만 override해보면 어떨까요??

jjj0611 commented on src/main/java/wooteco/subway/domain/Line.java:
```diff
@@ -0,0 +1,60 @@
+public class Line {
+
+    private final Long id;
+    private final String name;
+    private final String color;
```
이제 name이나 color는 null 값이 되어도 되는건가요??

bugoverdose replied:
validation 로직 추가했습니다

jjj0611 commented on src/main/java/wooteco/subway/domain/Station.java:
```diff
@@ -1,19 +1,19 @@
 
-    public Station() {
-    }
+    private final Long id;
+    private final String name;
```
ditto line에 남긴 코멘트와 함께 확인해주세요.

jjj0611 commented on src/main/java/wooteco/subway/service/LineService.java:
```diff
@@ -0,0 +1,48 @@
+        this.lineDao = lineDao;
+    }
+
+    public LineResponse save(Line line) {
+        return toLineResponse(lineDao.save(line));
+    }
```
save시에는 line domain을 사용하고 있고,
update 시에는 line request를 사용하고 있네요.

왜 두 방식에 차이점이 있나요??

bugoverdose replied:
로컬에서는 수정됐는데 리뷰용 커밋에서는 누락됐네요

jjj0611 commented on src/main/java/wooteco/subway/service/StationService.java:
```diff
@@ -0,0 +1,36 @@
+
+    private final StationDao stationDao;
+
+    public StationService(StationDao stationDao) {
+        this.stationDao = stationDao;
+    }
```
StationDao는 어디서 생성이 되며 언제 주입이 될까요??

jjj0611 replied:
이건 답변을 안하셨네요

bugoverdose replied:
IoC 컨테이너에서 생성해줍니다. IoC 컨테이너의 정체는 `ApplicationContext`의 구현체의 인스턴스 (굳이 명시하자면 그 상위 인터페이스인 `BeanFactory`의 구현체)로 알고 있습니다.

(1) StationDao 빈이 의존하는 빈들을 먼저 생성해주고, (2) 해당 빈들을 사용하여 StationDao 빈의 생성자를 호출하여 인스턴스화한 후, (3) StationDao 빈을 StationService의 생성자에 넣어줌으로써 StationService 빈을 생성하는 것으로 알고 있습니다.

jjj0611 commented on src/main/java/wooteco/subway/ui/LineController.java:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    @PostMapping
+    public ResponseEntity<LineResponse> createLine(@RequestBody LineRequest lineRequest) {
```
`@RequestBody` 어노테이션을 붙이면 어떤 일이 일어나나요?

jjj0611 replied:
이것도 답변을 안하셨네요

bugoverdose replied:
HTTP 요청 메시지의 바디에 담긴 JSON 데이터를 활용하여 LineRequest 인스턴스를 생성합니다. LineRequest에는 매개변수가 없는 디폴트 생성자와 각 필드별 setter가 호출되며, 만약 LineRequest의 필드들 중 JSON에 해당되는 정보가 없다면 null이 담기는걸로 기억합니다.

jjj0611 replied:
> 매개변수가 없는 디폴트 생성자와 각 필드별 setter가 호출되며

앞으로 계속해서 공부하시겠지만, 누가 그 역할을 해주는지, 꼭 setter만 가능한 것인지에 대한 것들을 공부해보면 좋을 것 같아요 :)

jjj0611 commented on src/test/java/wooteco/subway/dao/LineDaoTest.java:
```diff
@@ -0,0 +1,139 @@
+import wooteco.subway.exception.NotFoundException;
+
+@SuppressWarnings("NonAsciiCharacters")
+@SpringBootTest
```
어떤 일이 벌어지나요??

jjj0611 replied:
이것도 답변을 안하셨네요

bugoverdose replied:
`@SpringBootTest`을 붙이면 스프링 부트에서 메인 configuration 클래스(일반적으로 `@SpringBootApplication`가 붙은 클래스)를 찾아내고, 이를 기반으로 IoC 컨테이너 인스턴스를 생성하도록 합니다. IoC 컨테이너는 자동으로 @Controller 등의 어노테이션이 붙은 클래스들(그리고 해당 클래스들의 생성자 및 @Autowired가 붙은 필드 및 메서드들)을 참고하여 빈을 생성하고 의존성을 주입해줍니다.

`@SuppressWarnings("NonAsciiCharacters")`를 붙이면 테스트 메서드명으로 한글을 써도 경고가 뜨지 않습니다.

jjj0611 commented on src/test/java/wooteco/subway/dao/StationDaoTest.java:
```diff
@@ -0,0 +1,81 @@
+
+@SuppressWarnings("NonAsciiCharacters")
+@SpringBootTest
+@Transactional
```
제거되면 어떻게 되나요??
트랜잭션의 개념에 대해 알고 계실까요??

bugoverdose replied:
@Transactional을 붙이면 개별 테스트 메서드 실행 후, 그 결과를 테스트db에 커밋하지 않고 롤백해줍니다.
붙이지 않는 경우, findAll 테스트 후 save 테스트를 수행할 때 통과하던 테스트가, save 테스트 후 findAll 테스트를 수행하게 되면 실패하게 되겠죠.

jjj0611 replied:
엇 저는 `@Transaction`을 붙였을 경우 어떻게 작동하는지에 대한 질문을 한게 아니라
`트랜잭션`의 개념에 대해 아시는지 여쭤본거였어요

bugoverdose replied:
트랜잭션의 경우 (1) 하나 이상의 DML문으로 구성된 (2) 데이터베이스 조작 작업의 최소 단위라고 알고 있어요.

@Transactional은 기본적으로 복수의 DML 요청을 하나의 트랜잭션으로 묶어서 다룰 수 있도록 해준다. 트랜잭션이 지닌 특성들 중 원자성 때문에 모든 DML의 실행 결과는 함께 커밋되거나, 전부 롤백된다. 테스트의 경우 전부 롤백되고, 프로덕션의 경우 실패원자성을 지켜주기 때문에 추가해주면 좋다. 다만, Unchecked Exception이 발생할 경우 롤백되지 않을 수 있으므로 별도로 설정 추가해줘야 한다.

"트랜잭션의 개념"이라는 표현이 좀 광범위하네요. 어떤 점에 대해 더 설명하면 좋을까요.

jjj0611 replied:
엇 정 이정도면 충분해요
(2) 데이터베이스 조작 작업의 최소 단위 정도로만 이해하고 있어도 현 레벨에서는 충분하다고 생각합니다

jjj0611 commented on src/test/resources/acceptance_test_db.sql:
```diff
@@ -0,0 +1,17 @@
+DROP TABLE station IF EXISTS;
+DROP TABLE line IF EXISTS;
+
+CREATE TABLE station
```
acceptance_test_db와
dao_test_db를 따로 생성해두신 이유가 있을까요??

bugoverdose replied:
맨 위에서 설명했습니다. `schema.sql을 사용할 경우, 모든 테스트들이 하나의 test db에 종속될 것 같아, @Sql을 활용하여 테스트 관련 셋업/초기화 작업을 수행했습니다.`

모든 테스트들에 대해 동일한 테스트db를 셋업하여 사용하도록 하는 경우, INSERT문으로 사전에 저장해놓을 데이터를 자유롭게 수정하기 힘듭니다.
예를 들어 DAO 테스트를 위해 3개의 station 데이터를 4개로 수정했을 때, 인수테스트에서 "GET /stations"의 응답 메시지에 담기는 값도 굳이 수정해줘야 한다면 변경의 여파가 너무 크다고 볼 수 있죠. 그러니까 분리했습니다.

jjj0611 replied:
acceptance_test_db.sql을 사용하는 모든 테스트들이 동일한 셋업으로 테스트가 되는 것은 괜찮은건가요?
개별 테스트마다 필요한 데이터가 다를텐데, 그걸 sql 파일에 넣어두면 각 테스트마다 어떤 더미 데이터가 들어 있는지 확인하기 어렵지 않을까요?

bugoverdose replied:
음 "모든 테스트들이 동일한 셋업으로 테스트가 된다"고 했을 때 
(1) 테이블을 생성 및 초기화하는 부분(DDL)과 (2) INSERT로 더미 데이터를 추가하는 부분으로 구분된다고 생각하는데요.

(1)은 당연히 해야하는 작업이지만 "(2)의 경우 테스트 코드 내의 로직만으로는 DB에 어떤 데이터가 존재하는지 파악할 수 없다"는 점을 지적해주신 것 맞나요?

저도 이 점 때문에 인수 테스트 부분의 경우 sql 파일들에 INSERT문이 전혀 없고 DB를 생성하고 초기화하는 작업만 수행하였어요. (`acceptance_test_db.sql` 파일은 사라졌지만). 예를 들어 GET /lines에 대한 테스트 메서드를 보면 postLine이라는 private 메서드로 조회할 데이터를 명시적으로 저장하고 있습니다.

물론 dao 테스트들에서는 sql 파일에서 더미 데이터를 INSERT했는데, 그 이유는 DAO의 경우 '데이터베이스' 자체가 주요한 관심사라고 생각해서였습니다. 
예를 들어 만일 findById 에 대한 테스트에서 "dao의 save 메서드를 호출하여 데이터를 하나 생성하고, findById(1)을 했을 때 해당 데이터가 있을 것이다"라고 테스트한다면 이건 save에 대한 테스트인지, id 컬럼에 auto increment가 제대로 동작하는지를 테스트한 것인지, findById에 대한 테스트인지 모호해진다고 판단했어요.

이 부분은 트레이드 오프가 아닐까?라고 생각해요.

jjj0611 replied:
오 넵 정이 의도한 부분이 어떤 건지 이해되었어요.

그런데 findById(1)을 테스트하기 위해 미리 DB에 셋업한다면,
테스트 코드를 보고 해당 데이터가 어떻게 셋업이 되는지 알 방법이 없지 않을까요?

findAll이나 findBy(Page page)와 같은 메서드가 있을 때,
데이터가 얼마나 셋업되어 있는지를 테스트 코드만 봐서 파악하기는 힘들 것 같아요.

이 부분에 대해서 테스트를 셋업하는 것의 효용이 조금 와닿지는 않네요.
위에서 말한 것처럼 findById(1)을 테스트하기 위해 1개의 더미 데이터가 들어가 있어야 하고,
그 환경을 기준으로 save를 테스트하게 되면, save된 결과물의 id가 2가 나온다고 테스트를 해야하는데,
코드만 봐서는 왜 id의 검증이 2인지 알기 어려울 것 같아요

bugoverdose replied:
생각해보니 전 항상 아래처럼 좌우로 열어놓고 테스트를 작성해서 문제를 못 느꼈었는데, 어떤 점에서 문제기 될 수 있는지는 좀 이해는 되는 것 같아요. 예를 들어서 테스트 코드에서 실수로 잘못된 값을 입력한 경우, 그걸 찾아내는건 굉장히 힘들 수 있을 것 같다는 생각이 들긴하네요. 

<img width="1216" alt="image" src="https://user-images.githubusercontent.com/73531614/167831623-2bcafabc-8f0e-4943-bfa3-f9e0884cbdc8.png">

절충안에 대해 고민을 해보기는 해야할 것 같아요.

jjj0611 reviewed (changes requested):
안녕하세요 정! 리뷰어 제이입니다.

지하철 노선도 1, 2단계 구현하신 부분 잘 봤습니다.
전체적으로 간단하게 코멘트 남겨드렸으니 확인 부탁드려요.

이해가 안되는 부분이나 어려운 점이 있으시면 코멘트나 DM 부탁드려요!

bugoverdose committed:
refactor(controller): remove domain from LineController

- rename StationsService to StationService

bugoverdose committed:
feat(dao): implement StationDao2

- remove all validation logic from dao
- add test cases

bugoverdose committed:
feat(service): implement StationService2

- manage all validation logic at service
- add test cases

bugoverdose committed:
refactor(subway): replace legacy StationDao and StationService

- remove all unused code

bugoverdose committed:
feat(dao): implement LineDao2

- remove all validation logic from dao
- add test cases

bugoverdose committed:
feat(service): implement LineService2

- manage all validation logic at service
- add test cases

bugoverdose committed:
refactor(subway): replace legacy LineDao and LineService

- remove all unused code including StatementExecutor and UpdateResult

bugoverdose committed:
feat(subway): return status code 404 on data not found

- fix broken tests and README

bugoverdose committed:
refactor(test): fix and cleanse display names for controller tests

bugoverdose committed:
feat(test): return 400 for null or blank station name input

- add test cases

bugoverdose committed:
feat(test): return 400 for null or blank line name input

- add test cases

bugoverdose committed:
feat(service): add @Transactional for all CUD to ensure rollback on exception

bugoverdose committed:
refactor(domain): use only id field for equals/hashCode

- minor cleansing

bugoverdose committed:
test(acceptance): create single db and test instance for each test class

- remove @DirtiesContext and @Sql
- truncate and reset auto increment after each test
- reduce overall test time from 27 to 7 seconds

bugoverdose committed:
test(dao): remove meaningless @Transactional

- because @Sql resets the entire db after each test

jjj0611 reviewed (changes requested):
안녕하세요 정, 리뷰어 제이입니다.

리뷰요청 주신 것 확인했는데, 코멘트에 대한 답이 달려 있지 않은 것도 있고,
코멘트를 제대로 이해 못하신 것도 있는 것 같네요.

정이 코멘트 단 부분에는 추가로 답변 달아 두었고,
기존 코멘트에 답변이 제대로 되지 않은 부분이 있어 추가로 코드에 코멘트는 남기지 않았습니다.

답변 달아 둔 것 확인해보시고, 기존 코멘트에도 답변 다신 후 다시 요청 주세요~

bugoverdose committed:
refactor(dao): rename constants to CONSTANT_CASE format

jjj0611 reviewed (approved):
안녕하세요 정! 리뷰어 제이입니다.

컨벤션 수정하신거랑 코멘트 달아주신 것 확인하였습니다!

코드는 일단 3단계 진행을 위해 머지 할게요!

정이 남긴 코멘트의 답변은 오늘 내로 달아두도록 하겠습니다!
